core/src/metrics.cpp

```diff
@@ -0,0 +1,97 @@
+#include <set>
+#include <utility>
+
+#include <cubos/core/log.hpp>
+#include <cubos/core/metrics.hpp>
+#include <cubos/core/reflection/external/primitives.hpp>
+
+using cubos::core::Metric;
+using cubos::core::Metrics;
+using cubos::core::TimeTracker;
+
+namespace
+{
+    /// @brief Private type which stores the state of the metrics pool.
+    struct State
+    {
+        std::vector<Metric> mMetrics; ///< Collection of metrics.
+        std::set<std::string> mNames; ///< Available metric names.
+        mutable std::mutex mMutex;    ///< Mutex for thread-safe access to the metrics.
+    };
+} // namespace
+
+/// @brief Metrics pool state singleton. Guarantees it is initialized exactly once, when needed.
+/// @return Metrics pool.
+static State& state()
+{
+    static State instance{};
+    return instance;
+}
+
+void Metrics::metric(const Metric& metric)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+    state().mMetrics.emplace_back(metric);
+    state().mNames.insert(metric.mName);
+}
+
+size_t Metrics::size()
+{
+    return state().mMetrics.size();
+}
+
+bool Metrics::readValue(const std::string& name, double& value, size_t& offset)
+{
+    std::lock_guard<std::mutex> lock(state().mMutex);
+
+    std::vector<Metric> filteredMetricsByName;
```

Comment:
Why not store a map from metric names to vectors in the singleton?
This way you wouldnt need this expensive operation here